internal/interpreters/r.go

```diff
@@ -0,0 +1,378 @@
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/spf13/afero"
+
+	"github.com/posit-dev/publisher/internal/executor"
+	"github.com/posit-dev/publisher/internal/logging"
+	"github.com/posit-dev/publisher/internal/types"
+	"github.com/posit-dev/publisher/internal/util"
+)
+
+// Copyright (C) 2023 by Posit Software, PBC.
+
+const DefaultRenvLockfile = "renv.lock"
+
+var NotYetInitialized = types.NewAgentError(types.ErrorRExecNotFound, errors.New("not yet initialized"), nil)
+var MissingRError = types.NewAgentError(types.ErrorRExecNotFound, errors.New("unable to detect any R interpreters"), nil)
+var InvalidR = types.NewAgentError(types.ErrorRExecNotFound, errors.New("r executable is invalid"), nil)
+
+type RInterpreter interface {
+	Init() error
+
+	GetRExecutable() (util.AbsolutePath, error)
+	GetRVersion() (string, error)
+	GetLockFilePath() (util.RelativePath, bool, error)
+	CreateLockfile(util.AbsolutePath) error
+}
+
+type defaultRInterpreter struct {
+	executor   executor.Executor
+	pathLooker util.PathLooker
+
+	base                util.AbsolutePath
+	preferredPath       util.Path
+	rExecutable         util.AbsolutePath
+	version             string
+	lockfileRelPath     util.RelativePath
+	lockfileExists      bool
+	lockfileInitialized bool
+	log                 logging.Logger
+	initialized         bool
+	fs                  afero.Fs
+}
+
+var _ RInterpreter = &defaultRInterpreter{}
+
+var NewRInterpreter = RInterpreterFactory
+
+func RInterpreterFactory(base util.AbsolutePath,
+	rExecutableParam util.Path, log logging.Logger) RInterpreter {
+
+	return &defaultRInterpreter{
+		executor:   executor.NewExecutor(),
+		pathLooker: util.NewPathLooker(),
+
+		base:                base,
+		preferredPath:       rExecutableParam,
+		rExecutable:         util.AbsolutePath{},
+		version:             "",
+		lockfileRelPath:     util.RelativePath{},
+		lockfileExists:      false,
+		lockfileInitialized: false,
+		log:                 log,
+		initialized:         false,
+		fs:                  nil,
+	}
+}
+
+// Initializes the attributes within the defaultRInterpreter structure
+//  1. Resolves the path to the rExecutable to be used, with a preference
+//     towards the perferredPath, but otherwise, first one on path. If the
+//     executable is not a valid R interpreter, then will not be set.
+//  2. Seeds the version of the rExecutable being used, if set.
+//
+// We lazy load the lock file information, since it requires a call into
+// renv, which can be slow to be started (package initialization or something).
+// When this occurs, we'll do the following steps.
+//  3. Seeds the renv lock file for the rExecutable being used or if not found
+//     then the path to the default lock file
+//  4. Seeds the existance of the lock file at the lockfileRelPath
+//
+// Errors are taken care of internally and determine the setting or non-setting
+// of the attributes.
+func (i *defaultRInterpreter) Init() error {
```

Comment:
IMO, this should be part of the `constructor`, it can still be it's own method, but by running this logic in the constructor you'll get a couple goodies:
1. You won't have to check for `i.initialized` in every method.
2. Won't need to remember to run `Init()` after every new instance.

E.g:
```go
func NewRInterpreter(base util.AbsolutePath,
  rExecutableParam util.Path, log logging.Logger) (RInterpreter, error) {

  interpreter := &defaultRInterpreter{
    executor:   executor.NewExecutor(),
    // ...
  }
  err := interpreter.init()
  if err != nil {
    return nil, err
  }
  return interpreter, nil
}
```

Reply:
Yes, I had this in the original approach. Something that I can't recall right now drove me in this direction, but I'm not sure it is applicable anymore. I'll take another look and see if I have any concerns.

Reply:
OK, figured out that the reason I did this was so that we could update the structure assignment values ahead of the init function being run. This was allowing the unit testing to call the factory function, modify whatever they needed to do and then invoke init.

But, I've gone ahead and modified it to now use additional dependency injections for the functionality used internally. This caused some complications, but I do allow `nil` to be passed in and provide default providers for the functionality. So while there may be a few `nils` added to the inspect and interpreter factory methods (and perhaps a few more dependent factories), I think it is consistent.